src/connector/src/sink/file_sink/batching_log_sink.rs

```diff
@@ -0,0 +1,141 @@
+        loop {
+            let (epoch, item): (u64, LogStoreReadItem) = log_reader.next_item().await?;
+            if let LogStoreReadItem::UpdateVnodeBitmap(_) = &item {
+                match &state {
+                    LogConsumerState::BarrierReceived { .. } => {
+                        // we need to force to finish the batch here. Otherwise, there can be data loss because actor can be dropped and rebuilt during scaling.
+                        if let Some(committed_chunk_id) = sink_writer.should_finish().await? {
+                            log_reader.truncate(TruncateOffset::Chunk {
+                                epoch: (epoch),
+                                chunk_id: (committed_chunk_id),
+                            })?
+                        };
+                    }
+                    _ => unreachable!(
+                        "update vnode bitmap can be accepted only right after \
+                    barrier, but current state is {:?}",
+                        state
+                    ),
+                }
+            }
+            // begin_epoch when not previously began
+            state = match state {
+                LogConsumerState::Uninitialized => {
+                    // sink_writer.begin_epoch(epoch).await?;
+                    LogConsumerState::EpochBegun { curr_epoch: epoch }
+                }
+                LogConsumerState::EpochBegun { curr_epoch } => {
+                    assert!(
+                        epoch >= curr_epoch,
+                        "new epoch {} should not be below the current epoch {}",
+                        epoch,
+                        curr_epoch
+                    );
+                    LogConsumerState::EpochBegun { curr_epoch: epoch }
+                }
+                LogConsumerState::BarrierReceived { prev_epoch } => {
+                    assert!(
+                        epoch > prev_epoch,
+                        "new epoch {} should be greater than prev epoch {}",
+                        epoch,
+                        prev_epoch
+                    );
+                    // sink_writer.begin_epoch(epoch).await?;
+                    LogConsumerState::EpochBegun { curr_epoch: epoch }
+                }
+            };
+            match item {
+                LogStoreReadItem::StreamChunk { chunk, chunk_id } => {
+                    match sink_writer
+                        .write_batch_and_try_finish(chunk, chunk_id)
+                        .await
+                    {
+                        Err(e) => {
+                            return Err(e);
+                        }
+                        // The file has been successfully written and is now visible to downstream consumers.
+                        // Truncate the file to remove the specified `chunk_id` and any preceding content.
+                        Ok(Some(chunk_id)) => {
+                            log_reader.truncate(TruncateOffset::Chunk {
```

Comment:
Will there be any issue if we only do truncation by `TruncateOffset::Chunk` and never `TruncateOffset::Barrier`? Does log store rely on `TruncateOffset::Barrier`? cc @wenym1

Reply:
See https://github.com/risingwavelabs/risingwave/pull/18472/files#r1796507877